### ISSUE TITLE
Use "hasRole" instead of "hasAuthority" in SecurityConfiguration?

### ISSUE THREAD

**Issue by nickshoe** [jhipster/generator-jhipster#13994]
##### **Overview of the feature request**

I'm trying to understand the security model of spring-security, in particular I was looking for a clarification over the concepts of *authority* and *role*.

In [this article](https://www.baeldung.com/spring-security-granted-authority-vs-role) I've found the following explanation:
> ...in Spring Security, we can think of each Role as a coarse-grained GrantedAuthority that is represented as a String and prefixed with “ROLE“. When using a Role directly, such as through an expression like **hasRole(“ADMIN”)**, we are restricting access in a coarse-grained manner.

Since the generated project is not restricting resources in a fine-grained manner (it couldn't be otherwise, since it doesn't know the business requirements), I would expect it to use `hasRole(...)` instead of `hasAuthority(...)`.

This is one of the lines of code which could be affected by this change proposal:

https://github.com/jhipster/generator-jhipster/blob/5bdc374bcaf3553984524f19beaf90d1728aff43/generators/server/templates/src/main/java/package/config/SecurityConfiguration.java.ejs#L278

##### **Motivation for or Use Case**

The usage of `hasRole(...)` in the generated code would allow any new JHipster user to understand more easily the security model behind the scenes.

##### **Related issues or PR**

I can do the PR if the change proposal makes sense.

- [x ] Checking this box is mandatory (this is just to show you read everything)

**Comment by nickshoe** [comment]
From the point of view of Spring Security framework, there is [no difference](https://stackoverflow.com/questions/19525380/difference-between-role-and-grantedauthority-in-spring-security) between _roles_ and _authorities_ when it comes to authorizing a request (apart from the former using the prefix "ROLE_" and its implications in the library code and the client code).
Both actually are **opaque** strings that are searched for in the autheticated user's _GrantedAuthorities_ list, when the framework needs to detemine if the user is authorized to perform an action or not; they don't carry any semantic value.

https://github.com/spring-projects/spring-security/blob/857830f6958c229650e6a00d376af38c8bcddbad/core/src/main/java/org/springframework/security/access/expression/SecurityExpressionRoot.java#L86-L89

https://github.com/spring-projects/spring-security/blob/857830f6958c229650e6a00d376af38c8bcddbad/core/src/main/java/org/springframework/security/access/expression/SecurityExpressionRoot.java#L96-L99

Anyways, since the framework supports these two concepts and there actually is a distinction between the two from the semantic point of view (i.e. _roles_ are intended as coarse-grained permissions or, better, containers of several authorities; _authorities_ are fine-grained permissions, possibly related to roles in, respectively, a child-parent relationship) I would suggest to use _roles_ for the generator so as to better "guide" the new jhipster's users towards the right tool for their security requirements, and avoiding semantic confusion.

**Comment by gmarziou** [comment]
@nickshoe what you propose then it means also we should rename the Angular directives (and their equivalent in React and Vue) and adjust the HTML templates that use them. It's a big impact on existing projects that would want to upgrade their JHipster generator.

Also, existing articles, books or answers on Stackoverflow would be obsolete or at least confusing to new users.

So, I'm not convinced the result of this change would be positive.

**Comment by mraible** [comment]
After many years of using both, I think "hasAuthority" is better.

**Comment by nickshoe** [comment]
> @nickshoe what you propose means also that, for consistency, we should rename the Angular directives (and their equivalent in React and Vue) and adjust the HTML templates that use them. It's a big impact on existing projects that would want to upgrade their JHipster generator.
> 
> Also, existing articles, books or answers on Stackoverflow would be obsolete or at least confusing to new users.
> 
> So, I'm not convinced the result of this change would be positive.

I understand, but if this change makes sense then it could be considered a breaking change. Backward compatibility is important, but I think It shouldn't prevent enhancements.

I'm not afraid ti address all the changes needed in the generator (backend and frontend), I can do the PR.

The point is if using `hasRole` makes any sense and if it is considered a good practice.

**Comment by nickshoe** [comment]
@mraible could you expand on this a bit? 
Is there a particular reason why is not convenient to use `hasRole` and can you give an example?

**Comment by mraible** [comment]
`hasRole()` has been a source of headache and pain on many projects. Both in the open-source world and when I was a consultant several years ago. `hasAuthority()` works as expected without surprises.

**Comment by nickshoe** [comment]
Since I don't have enough experience to argue that `hasRole` is a better implementation choice compared to `hasAuthority` (my point was mainly to shed a light on the semantics of roles and authorities and a possibile implementation), I would ask a third party to take a decision.

What I've learned: authorities are enough when It comes to decide whether authorizing somebody or not to do something; in this case we are not interested in the hierarchical aspect between roles and authorities, this is a business model concern the may be implemented upstream to load the authorities (starting from role(s)) for the principal.
In other words: when you see your boss doing something you can't do, you may just think he can do that because he/she is your boss (role) instead of thinking of all the specific permissions (authorities) he/she gas.

**Comment by nickshoe** [comment]
If we want to only use "authorities", then I think we should rename the following method to something like `hasCurrentUserThisAuthority`:

https://github.com/jhipster/generator-jhipster/blob/5bdc374bcaf3553984524f19beaf90d1728aff43/generators/server/templates/src/main/java/package/security/SecurityUtils.java.ejs#L147

**Comment by mraible** [comment]
I'm OK with that. Can you please create a PR with this change?

**Comment by nickshoe** [comment]
Sure!

**Comment by nickshoe** [comment]
Done. 

[This](https://github.com/jhipster/generator-jhipster/pull/14147/checks?check_run_id=2024540143) test is failing, but maybe it was due to perfomance issues resulting in timeouts.

**Comment by mraible** [comment]
@nickshoe I've restarted the tests to see if it helps.

**Comment by nickshoe** [comment]
@nickshoe now everything looks good. 
Thank you
